fix(docs-infra): include correct dependencies in StackBlitz examples

gkalpak opened the pull request:
Previously, all StackBlitz examples included the default dependencies for `cli`-type projects. However, different example types may have different `package.json` files with different dependencies.
For example, the [boilerplate `package.json`][1] for `elements` examples includes an extra dependency on `@angular/elements`.

This commit changes `StackblitzBuilder` to use the dependencies that correspond to each example type.
(NOTE: Manually verified the changes.)

Jira issue: [FW-2002][2]

[1]: https://github.com/angular/angular/blob/05d058622/aio/tools/examples/shared/boilerplate/elements/package.json
[2]: https://angular-team.atlassian.net/browse/FW-2002

[FW-2002]: https://angular-team.atlassian.net/browse/FW-2002

mary-poppins commented:
You can preview 3de2d7a at https://pr36071-3de2d7a.ngbuilds.io/.

mary-poppins commented:
You can preview ea135b1 at https://pr36071-ea135b1.ngbuilds.io/.

mary-poppins commented:
You can preview 4f6f279 at https://pr36071-4f6f279.ngbuilds.io/.

petebacondarwin commented on aio/tools/stackblitz-builder/builder.js:
```diff
@@ -68,30 +67,29 @@ class StackblitzBuilder {
-      var partPath = path.dirname(path.relative(this.basePath, outputFileName));
-      var altFileName = path.join(this.destPath, partPath, path.basename(outputFileName)).replace('.no-link.', '.');
+      const partPath = path.dirname(path.relative(this.basePath, outputFileName));
+      altFileName = path.join(this.destPath, partPath, path.basename(outputFileName)).replace('.no-link.', '.');
```
👍

petebacondarwin commented on aio/tools/stackblitz-builder/builder.js:
```diff
@@ -118,85 +116,87 @@ class StackblitzBuilder {
+          document.getElementById("mainForm").submit();
+        </script>
+      </body></html>
+    `.trim();
```
Is the `trim()` really needed?

gkalpak replied:
Probably not, but added it just in case :grin:

petebacondarwin commented on aio/tools/stackblitz-builder/builder.js:
```diff
@@ -118,85 +116,87 @@ class StackblitzBuilder {
+      if (extn === '.png') {
         content = this._encodeBase64(fileName);
-        fileName = fileName.substr(0, fileName.length - 4) + '.base64.png'
+        fileName = `${fileName.slice(0, -extn.length)}.base64${extn}`;
```
🤓

petebacondarwin commented on aio/tools/stackblitz-builder/builder.js:
```diff
@@ -208,65 +208,56 @@ class StackblitzBuilder {
-    tags.forEach(function(tag,ix) {
-      postData['tags[' + ix + ']'] = tag;
-    });
+    const tags = ['angular', 'example'].concat(config.tags || []);
```
Perhaps:
```suggestion
    const tags = ['angular', 'example', ...config.tags || []];
```

petebacondarwin commented on aio/tools/stackblitz-builder/builder.js:
```diff
@@ -208,65 +208,56 @@ class StackblitzBuilder {
+    const doc = jsdom.jsdom(baseHtml);
+    const form = doc.querySelector('form');
+
+    Object.entries(postData).forEach(([key, value]) => {
```
We can now also do:

```suggestion
    for(const [key, value] of postData) {
```

gkalpak replied:
You mean `... of Object.entries(postData)`, right?
We can indeed. The question is, should we? 😛
<sub>(I did anyway - just because I can 😛 </sub>

petebacondarwin replied:
No! What I wrote works in node 12.14.1 as-is.

gkalpak replied:
On 12.15.0, I get `postData is not iterable` 😕

petebacondarwin commented on aio/tools/stackblitz-builder/builder.js:
```diff
@@ -208,65 +208,56 @@ class StackblitzBuilder {
+
     try {
-      var config = (configSrc && configSrc.trim().length) ? JSON.parse(configSrc) : {};
+      config = (configSrc && configSrc.trim().length) ? JSON.parse(configSrc) : {};
```
It's a shame that JS doesn't have a good way of initializing a const variable inside a try-catch block.

petebacondarwin replied:
I guess the simplest way would be to put it in a function and call it:

```js
const config = parseConfig(configSrc);
```

petebacondarwin reviewed (approved):
LGTM - nothing blocking

gkalpak committed:
refactor(docs-infra): clean up `stackblitz-builder/builder.js` script

- Remove unused dependencies.
- Change `var` to `const/let`.
- Change regular functions as callbacks to arrow functions.
- Remove unnecessary intermediate variables.
- Switch from custom `_existsSync()` implementation to built-in
  `fs.existsSync()`.

gkalpak committed:
fix(docs-infra): include correct dependencies in StackBlitz examples

Previously, all StackBlitz examples included the default dependencies
for `cli`-type projects. However, different example types may have
different `package.json` files with different dependencies.
For example, the [boilerplate `package.json`][1] for `elements` examples
includes an extra dependency on `@angular/elements`.

This commit changes `StackblitzBuilder` to use the dependencies that
correspond to each example type.
(NOTE: Manually verified the changes.)

Jira issue: [FW-2002][2]

[1]: https://github.com/angular/angular/blob/05d058622/aio/tools/examples/shared/boilerplate/elements/package.json
[2]: https://angular-team.atlassian.net/browse/FW-2002

gkalpak committed:
fixup! refactor(docs-infra): clean up `stackblitz-builder/builder.js` script

gkalpak commented:
Thx for reviewing, @petebacondarwin ✨
Rebased on master and addressed comments. Marking for merge...

mary-poppins commented:
You can preview 7174f35 at https://pr36071-7174f35.ngbuilds.io/.

mhevery commented:
MERGE_ASSISTANCE: the PR is ready for merge but the bot seems stuck.
